libevm/VMSIMD.cpp

```diff
@@ -0,0 +1,726 @@
+void VM::xslt (uint8_t _type) { EVALXOPS(LT,  _type); }
+void VM::xgt  (uint8_t _type) { EVALXOPU(GT,  _type); }
+void VM::xsgt (uint8_t _type) { EVALXOPS(GT,  _type); }
+void VM::xeq  (uint8_t _type) { EVALXOPU(EQ,  _type); }
+void VM::xzero(uint8_t _type) { EVALXOPU(ZERO,_type); }
+void VM::xand (uint8_t _type) { EVALXOPU(AND, _type); }
+void VM::xoor (uint8_t _type) { EVALXOPU(OR,  _type); }
+void VM::xxor (uint8_t _type) { EVALXOPU(XOR, _type); }
+void VM::xnot (uint8_t _type) { EVALXOPU(NOT, _type); }
+void VM::xshr (uint8_t _type) { EVALXOPU(SHR, _type); }
+void VM::xsar (uint8_t _type) { EVALXOPS(SHR, _type); }
+void VM::xshl (uint8_t _type) { EVALXOPU(SHL, _type); }
+void VM::xrol (uint8_t _type) { EVALXOPU(ROL, _type); }
+void VM::xror (uint8_t _type) { EVALXOPU(ROR, _type); }
+
+inline uint8_t pow2N(uint8_t _n)
+{
+	static uint8_t exp[6] = { 1, 2, 4, 8, 16, 32 };
+	return exp[_n];
+}
+
+inline uint8_t laneCount(uint8_t _type)
+{
+	return pow2N((_type) & 0xf);
+}
+
+inline uint8_t laneWidth(uint8_t _type)
+{
+	return (_type) >> 4;
+}
+
+// in must be by reference because it is really just memory for a vector
+u256 VM::vtow(uint8_t _type, const u256& _in)
+{
+	u256 out;
+	uint8_t const count = laneCount(_type);
+	uint8_t const width = laneWidth(_type);
+	switch (width)
+	{
+	case Bits8:
+		for (int i = count - 1; 0 <= i; --i)
+		{ 
+			out << 8;
+			out |= v8x32(_in) [i];
+		}
+		break;
+	case Bits16:
+		for (int i = count - 1; 0 <= i; --i)
+		{ 
+			out << 16;
+			out |= v16x16(_in)[i];
+		}
+		break;
+	case Bits32:
+		for (int i = count - 1; 0 <= i; --i)
+		{ 
+			out << 32;
+			out |= v32x8(_in) [i];
+		}
+		break;
+	case Bits64:
+		for (int i = count - 1; 0 <= i; --i)
+		{ 
+			out << 64;
+			out |= v64x4(_in) [i];
+		}
+		break;
+	default:
+		throwBadInstruction();
+	}
+	return out;
+}
+
+// out must be by reference because it is really just memory for a vector
+void VM::wtov(uint8_t _type, u256 _in, u256& o_out)
+{
+	uint8_t const count = laneCount(_type);
+	uint8_t const width = laneWidth(_type);
+	switch (width)
+	{
+	case Bits8:
+		for (int i = count - 1; 0 <= i; --i)
+		{ 
+			v8x32(o_out) [i] = (uint8_t )(_in & 0xff);
+			_in >>= 8;
+		}
+		break;
+	case Bits16:
+		for (int i = count - 1; 0 <= i; --i)
+		{ 
+			v16x16(o_out)[i] = (uint16_t)(_in & 0xffff);
+			_in >>= 16;
+		}
+		break;
+	case Bits32:
+		for (int i = count - 1; 0 <= i; --i)
+		{ 
+			v32x8(o_out) [i] = (uint32_t)(_in & 0xffffff);
+			_in >>= 32;
+		}
+		break;
+	case Bits64:
+		for (int i = count - 1; 0 <= i; --i)
+		{ 
+			v64x4(o_out) [i] = (uint64_t)(_in & 0xffffffff);
+			_in >>= 64;
+		}
+		break;
+	default:
+		throwBadInstruction();
+	}
+}
+
+void VM::xmload (uint8_t _type)
+{
+	// goes onto stack element by element, LSB first
+	uint8_t const* p = m_mem.data() + toInt15(m_SP[0]);
+	uint8_t const count = laneCount(_type);
+	uint8_t const width = laneWidth(_type);
+
+	switch (width)
+	{
+	case Bits8:
+		for (int j = 1,  i = count - 1; 0 <= i; --i)
```

Comment:
Is `j` initialization correct here and in other cases? it will go negative here if count > 2

It was initializeed as `j = n` in earlier commits